### PR TITLE
Lab 02- Sort algorithm implementation

### PR DESCRIPTION
**Pull request by jecofang01**
Fixes #42

#### Changes proposed in this PR:

* Implement bubble sort
* Implement insertion sort
* Implement quick sort
* Add test cases
* Fix the comments

### CODE REVIEW

File: 02_bubble/jecofang/sort.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	if n < 2 {
+		return s
+	}
+
+	swap := true
+	for swap {
+		swap = false
+		for i := 0; i < n-1; i++ {
+			if s[i+1] < s[i] {
+				s[i], s[i+1] = s[i+1], s[i]
```
Comment (juliaogris) [inline 02_bubble/jecofang/sort.go]: Make copy of s before modifying (see lab problem statement).

**Review by juliaogris — CHANGES_REQUESTED** [review]
Again, very solid, the only update request: please make a copy of the array to be sorted first as per problem statement (you will otherwise modify the input).
Please also add to tests checks for unmodified input.

File: 02_bubble/jecofang/sort.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+func quick(s []int) []int {
+	n := len(s)
+
+	if n < 2 {
+		arr := make([]int, n)
+		copy(arr, s)
+		return arr
+	}
+
+	left := make([]int, 0, n)
```
Comment (linwumeng) [inline 02_bubble/jecofang/sort.go:63]: use many memory here. Try swapping on the slice.

File: 02_bubble/jecofang/sort_test.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+var testData = []struct {
+	in  []int
+	out []int
+}{
+	{[]int{3, 2, 1, 5}, []int{1, 2, 3, 5}},
+	{[]int{3, 2, 1, 5, -1}, []int{-1, 1, 2, 3, 5}},
+	{[]int{3, 2, 1, 5, -1, 3}, []int{-1, 1, 2, 3, 3, 5}},
+	{[]int{}, []int{}},
```
Comment (linwumeng) [inline 02_bubble/jecofang/sort_test.go:32]: test:
0 element
1 element
2 elements
3 elements
10 elements
all positive
all negative
all 0
sorted data
inverse sorted data

File: 02_bubble/jecofang/sort.go (hunk 2 above)
Comment (juliaogris) [reply 02_bubble/jecofang/sort.go:63]: Good point, @linwumeng - but then @jecofang01 you should probably add a comment "sorts []int in place" - or you need a wrapping function that first creates a copy or the original array.

**Review by juliaogris — APPROVED** [review]
As far as I can tell you seem like a really talented and diligent coder to me - don't you think @anzdaddy ? 
I hope you enjoy your job!

LGTM 👍

**Pull request merged**